### PR TITLE
Authentication integration into server side and front end

### PR DESCRIPTION
**Pull request by erikrramsey**
**(Sending requests for review to the contributors whose code I modified)**

**Changes will be deployed on the server after this PR is accepted, and then I will make a second PR that includes the deployment endpoints**

Initial creation of server-side, local (on our database) authentication.

Additional example usage, and utility functions included in the apps front end.

New dependencies:
Carbon:
* react-native-async-storage

CarbonServer:
* passport
* passport-jwt
* passport-local
* jsonwebtoken
* bcrypt

Changes:
Server:
* Added auth.js to define the strategies needed to utilize passport authentication.
* auth.js handles decrypting JWT's, hashing user passwords on signup, and most importantly authenticating a login.
* Added a new route to user.js: user/testrank that uses a jwt instead of a raw user id
* Commented out user/:id and user/rank in order to make use of the new routes.
* Refactored user (model) to User. Matches capitalization of other similar models.

Carbon:
LoginScreen:
* Modified LoginScreen to add login and signup functionality.
(Currently there is no forgot password or account registration yet, as we don't have an email service set up.)

MainContainer:
* Changed MainContainer to take you directly to the loginscreen
* LoginScreen is in a stack nav with the SignupScreen as well now
* In order to use app you must be signed in as a user

LoginManager:
* Added LoginManager in new folder called util
* Has utility functions for authentication getToken(), getAuthHeader(), login, signup, logout
* Use getAuthHeader() in fetch calls that utilize jwt's

RankingScreen:
* User rank is now fetched using the new enpoint that utilizes JWT's
* This was intended to set an example of how simple it is to change an endpoint to use the getAuthHeader() function.

### CODE REVIEW

**Review by mcebaniqued — APPROVED** [review]

File: Carbon/util/LoginManager.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        var formBody = [];
+        for (var property in details) {
+            var encodedKey = encodeURIComponent(property);
+            var encodedValue = encodeURIComponent(details[property]);
+            formBody.push(encodedKey + "=" + encodedValue);
+        }
+        formBody = formBody.join("&");
+
+        var response = await fetch('http://192.168.0.150:3000/api/user/auth/login',
```
Comment (aq6476) [inline Carbon/util/LoginManager.js:37]: Can this be changed to the config/api export? I think maybe for each link we just have them all declared there for specific routes.

File: Carbon/util/LoginManager.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    var formBody = [];
+    for (var property in details) {
+        var encodedKey = encodeURIComponent(property);
+        var encodedValue = encodeURIComponent(details[property]);
+        formBody.push(encodedKey + "=" + encodedValue);
+    }
+    formBody = formBody.join("&");
+
+    var response = await fetch('http://192.168.0.150:3000/api/user/auth/signup',
```
Comment (aq6476) [inline Carbon/util/LoginManager.js:81]: Same issue as mentioned earlier in the file.

**Review by aq6476 — CHANGES_REQUESTED** [review]
Hey Erik. Firstly, great feature!

I was running this branch on my end and found no issues, everything seems to work as intended for the authentication. Thank you for taking the helm on getting authentication and user sessions implemented. Theres not much to criticize, the feature was added as a modular component, so it doesn't conflict with anything else in the project, which is great! I posted these two comments earlier but in  Carbon/Util/LoginManager.js can you change it to the export from Carbon/config/Api.js ? that way we don't have to dig in when changing on our end. Other than that, great work. After those changes i will approve the PR : )

**Review by JaneenSudiacal — APPROVED** [review]
Good job on this PR! I think it looks great and easy to understand. I really appreciate you adding functionality and navigation to the sign up and login pages I designed. Approved!

File: Carbon/util/LoginManager.js (hunk 2 above)
Comment (erikrramsey) [reply Carbon/util/LoginManager.js]: Should be fixed

**Review by aq6476 — APPROVED** [review]
With the API URL changed, I'm approving this PR it looks good to deploy 👍

**Pull request merged**